pkg/koordlet/util/blkio_scanner.go

```diff
@@ -0,0 +1,357 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog/v2"
+)
+
+// disk refers to a physical storage device such as a hard disk, solid-state disk
+// disk number is major:minor of disk
+// part is short for partition
+// device here contains only disks and partitions
+type LocalStorageScanner interface {
+	GetDiskNumber(disk string) string
+	GetDiskByNumber(number string) string
+	GetDiskByDevice(device string) string
+	GetDiskByVG(vgName string) string
+	GetVGByLV(lvName string) string
+	GetDiskByMountPoint(mountpoint string) string
+	Run(stopCh <-chan struct{})
+	Scan()
+	IsAvailable() bool
+}
+
+type localStorageScanner struct {
+	// map is not thread safe. so "concurrent read/write of Map" can cause error.
+	rwLock sync.RWMutex
+	// mapper of disk and disk number, such as "/dev/vda":"253:0"
+	diskNumberMap map[string]string
+	// mapper of disk number and disk, such as "253:0":"/dev/vda"
+	numberDiskMap map[string]string
+	// mapper of partition and its disk, such as "/dev/vdb3":"/dev/vdb"
+	partitionDiskMap map[string]string
+	// mapper of volumegroup and its disk, such as "yoda-pool0":"/dev/vdb"
+	vgDiskMap map[string]string
+	// mapper of logicalvolume and its volumegroup, such as "/dev/mapper/yoda--pool0-yoda--2c52d97f--eab6--4ac5--ba8b--242f399470e1":"yoda-pool0"
+	lvMapperVGMap map[string]string
+	// mapper of mountpoint and its disk, such as "/var/lib/kubelet/pods/d806ee8d-fe28-4995-a836-d2356d44ec5f/volumes/kubernetes.io~csi/yoda-2c52d97f-eab6-4ac5-ba8b-242f399470e1/mount":"/dev/mapper/yoda--pool0-yoda--2c52d97f--eab6--4ac5--ba8b--242f399470e1"
+	mpDiskMap    map[string]string
+	scanInterval time.Duration
+}
+
+var (
+	keyCommands   = []string{"lsblk", "vgs", "lvs", "findmnt"}
+	deviceName    = "/dev/%s"
+	lvmMapperName = "/dev/mapper/%s-%s"
+
+	lsblkRE   = regexp.MustCompile(`([A-Z:]+)=(?:"(.*?)")`)
+	vgsRE     = regexp.MustCompile(`[^\s]+`)
+	lvsRE     = regexp.MustCompile(`[^\s]+`)
+	findmntRE = regexp.MustCompile(`([A-Z:]+)=(?:"(.*?)")`)
+
+	lsblkColumns = []string{
+		"NAME",
+		"TYPE",
+		"MAJ:MIN",
+	}
+	vgsColumns = []string{
+		"vg_name",
+		"pv_count",
+		"pv_name",
+	}
+	lvsColumns = []string{
+		"lv_name",
+		"vg_name",
+	}
+	findmntColumns = []string{
+		"TARGET",
+		"SOURCE",
+	}
+)
+
+func NewLocalStorageScanner(interval time.Duration) LocalStorageScanner {
```

Comment:
could we move the scanner as a plugin of the metric advisor(collector)?

Reply:
done